File: problem-solving/_lc-28-easy-find-the-index-of-the-first-occurrence-in-a-string/1-solution.py
"""
https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/

Complexity
----------
Time: O(m*n)
Space: O(m*n)

NOTE
----
This can be solved in linear time using standard pattern matching algorithms like
KMP, Rabin karp, ...
"""

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        m, n = len(haystack), len(needle)
        for i in range(m - n + 1):
            for j in range(n):
                if haystack[i + j] != needle[j]:
                    break
                if j == n - 1:
                    return i
        return -1

File: problem-solving/lc-274-medium-h-index/1-solution.go
/*
https://leetcode.com/problems/h-index/

Complexity
----------
Time: O(n)
Space: O(n)
*/

func hIndex(citations []int) int {
    n := len(citations)
    buckets := make([]int, n + 1)
    for _, c := range citations {
        if c >= n {
            buckets[n] += 1
        } else {
            buckets[c] += 1
        }
    }

    total := 0
    for i := n; i >= 0; i-- {
        total += buckets[i]
        if total >= i {
            return i
        }
    }
    return 0
}
